src/license_guc.c

```diff
@@ -71,315 +43,218 @@ current_license_can_downgrade_to_apache(void)
-	return false;
-}
-
-/* SQL functions */
-
-PGDLLEXPORT Datum
-ts_tsl_loaded(PG_FUNCTION_ARGS)
-{
-	if (TS_CURRENT_LICENSE_IS_APACHE_ONLY())
-		PG_RETURN_BOOL(false);
-
-	PG_RETURN_BOOL(ts_cm_functions->check_tsl_loaded());
 }
 
-PGDLLEXPORT Datum
-ts_enterprise_enabled(PG_FUNCTION_ARGS)
+static void
+tsl_module_init(void)
 {
-	if (TS_CURRENT_LICENSE_IS_APACHE_ONLY())
-		PG_RETURN_BOOL(false);
-
-	PG_RETURN_BOOL(ts_cm_functions->enterprise_enabled_internal());
-}
-
-PGDLLEXPORT Datum
-ts_current_license_key(PG_FUNCTION_ARGS)
-{
-	if (!is_member_of_role(GetUserId(), DEFAULT_ROLE_READ_ALL_SETTINGS))
-		ereport(ERROR,
-				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-				 errmsg("must be superuser or a member of pg_read_all_settings to examine the "
-						"license key")));
-	Assert(ts_guc_license_key != NULL);
-	PG_RETURN_TEXT_P(cstring_to_text(ts_guc_license_key));
-}
-
-PGDLLEXPORT Datum
-ts_print_tsl_license_expiration_info(PG_FUNCTION_ARGS)
-{
-	if (ts_cm_functions->print_tsl_license_expiration_info_hook != NULL)
-		ts_cm_functions->print_tsl_license_expiration_info_hook();
-	PG_RETURN_VOID();
+	Assert(tsl_handle != NULL);
+	Assert(tsl_init_fn != NULL);
+	DirectFunctionCall1(tsl_init_fn, CharGetDatum(0));
 }
 
-PGDLLEXPORT Datum
-ts_license_expiration_time(PG_FUNCTION_ARGS)
+static void
+tsl_module_shutdown(void)
 {
-	if (ts_cm_functions->print_tsl_license_expiration_info_hook == NULL)
-		PG_RETURN_TIMESTAMPTZ(DT_NOEND);
-
-	PG_RETURN_TIMESTAMPTZ(ts_cm_functions->license_end_time());
+	if (ts_cm_functions->module_shutdown_hook != NULL)
+		ts_cm_functions->module_shutdown_hook();
 }
 
-PGDLLEXPORT Datum
-ts_license_edition(PG_FUNCTION_ARGS)
+/*
+ * Check hook function set by license guc.
+ *
+ * Used to validate license string before the assign hook
+ * ts_license_guc_assign_hook() call.
+ */
+bool
+ts_license_guc_check_hook(char **newval, void **extra, GucSource source)
 {
-	char *edition = NULL;
+	LicenseType type = license_type_of(*newval);
 
-	switch (TS_CURRENT_LICENSE_TYPE())
+	switch (type)
 	{
-		case LICENSE_TYPE_APACHE_ONLY:
-			edition = TS_LICENSE_APACHE_ONLY_PRINT_TEXT;
-			break;
-		case LICENSE_TYPE_COMMUNITY:
-			edition = TS_LICENSE_COMMUNITY_PRINT_TEXT;
+		case LICENSE_APACHE:
+			/* Allow setting Apache license only if is is set from postgresql.conf
+			 * or the server command line */
+			if (source == PGC_S_FILE || source == PGC_S_ARGV)
+				break;
+			if (!license_is_downgradable())
+			{
+				GUC_check_errdetail("Cannot downgrade a running session to Apache license.");
+				GUC_check_errhint("Change the license in the configuration file.");
+				return false;
+			}
 			break;
-		case LICENSE_TYPE_ENTERPRISE:
-			edition = TS_LICENSE_ENTERPRISE_PRINT_TEXT;
+		case LICENSE_TSL:
 			break;
-		default:
-			elog(ERROR, "Invalid license key '%s'", ts_guc_license_key);
-			pg_unreachable();
+		case LICENSE_UNDEF:
+			return false;
+	}
+
+	if (!load_enabled)
+	{
+		load_source = source;
+		return true;
```

Comment:
Why do we return `true` here even though load is not enabled? Feels like it should be a failure, or is the loading deferred somehow?

I guess I am wondering what the difference is between returning `true` here and returning `true` below (at the end of the function).

Reply:
I think this is to allow delayed loading to work. Basically it saves the GucSource value and then retries it later